File: clients/omni-windows/src/Omni.Ipc/OmniPaths.cs
using System.Security.Cryptography;
using System.Text;

namespace Omni.Ipc;

/// <summary>
/// Locates the daemon's state directory and named pipe, reproducing the daemon's
/// own derivation (see <c>crates/omni-runtime/src/config.rs</c>) so a non-Rust
/// client can find it without the daemon publishing an address.
/// </summary>
public static class OmniPaths
{
    /// <summary>
    /// The daemon's state directory: <c>%OMNI_CONFIG_DIR%</c> if set, otherwise
    /// <c>%APPDATA%\omni</c> (the platform config dir, matching Rust's
    /// <c>dirs::config_dir()</c> on Windows).
    /// </summary>
    public static string ConfigDir()
    {
        var overridden = Environment.GetEnvironmentVariable("OMNI_CONFIG_DIR");
        if (!string.IsNullOrEmpty(overridden))
        {
            return overridden;
        }

        var appData = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
        return Path.Combine(appData, "omni");
    }

    /// <summary>
    /// The short pipe name (without the <c>\\.\pipe\</c> prefix) for a state
    /// directory, as <see cref="System.IO.Pipes.NamedPipeClientStream"/> expects
    /// it. Derived from a stable SHA-256 of the directory path so two daemons with
    /// different <c>OMNI_CONFIG_DIR</c> get distinct pipes — and so this client and
    /// the Rust daemon agree on the name.
    /// </summary>
    public static string PipeShortName(string? configDir = null)
    {
        configDir ??= ConfigDir();
        var digest = SHA256.HashData(Encoding.UTF8.GetBytes(configDir));
        var hex = Convert.ToHexStringLower(digest.AsSpan(0, 8));
        return $"omni-{hex}";
    }

    /// <summary>The full kernel pipe path, for display and diagnostics.</summary>
    public static string PipeName(string? configDir = null) =>
        $@"\\.\pipe\{PipeShortName(configDir)}";
}

File: clients/omni-macos/omni-macos/IPC/OmniPaths.swift
//
//  OmniPaths.swift
//  omni-macos
//
//  Locates the daemon's Unix-domain socket, reproducing the daemon's own
//  derivation (see `crates/omni-runtime/src/config.rs`) so this client finds it
//  without the daemon publishing an address.
//

import Foundation

public nonisolated enum OmniPaths {
    /// The daemon's state directory: `$OMNI_CONFIG_DIR` if set, otherwise the
    /// platform config directory with `omni` appended. On macOS that is
    /// `~/Library/Application Support/omni`, matching Rust's
    /// `dirs::config_dir()` (which maps to Application Support).
    public static func configDirectory(
        environment: [String: String] = ProcessInfo.processInfo.environment
    ) -> URL {
        if let override = environment["OMNI_CONFIG_DIR"], !override.isEmpty {
            return URL(fileURLWithPath: override, isDirectory: true)
        }
        let base = FileManager.default
            .urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? URL(fileURLWithPath: NSHomeDirectory())
                .appendingPathComponent("Library/Application Support", isDirectory: true)
        return base.appendingPathComponent("omni", isDirectory: true)
    }

    /// The Unix-domain socket the daemon listens on (mode 0600): `daemon.sock`
    /// inside the state directory.
    public static func socketPath(
        environment: [String: String] = ProcessInfo.processInfo.environment
    ) -> String {
        configDirectory(environment: environment)
            .appendingPathComponent("daemon.sock", isDirectory: false)
            .path
    }
}
